newsroom/news_api/api_tokens/auth.py

```diff
@@ -0,0 +1,139 @@
+            403, gettext("Authorization token missing."), title=gettext("403. Forbidden")
+        )
+        token_id = self.get_token_from_request(request)
+        if token_id is None:
+            raise token_missing_exception
+
+        # TODO-ASYNC: replace when api_tokens are async
+        token = tokens_service.find_one(req=None, _id=token_id)
+        if token is None:
+            raise token_missing_exception
+
+        await self.check_token_validity(request, token)
+        await self.start_session(request, token)
+        await self.check_rate_limit(request, token)
+
+    async def start_session(self, request: Request, token: dict[str, Any]):
+        company = await CompanyService().find_by_id(token.get("company"))
+
+        # TODO-ASYNC: replace token with actual token resource model once api_tokens is async
+        request.storage.request.set("company_auth_token", token)
+        request.storage.request.set("company_instance", company)
+
+        # TODO-ASYNC: check if we need this here or refactor it with request.storage
+        g.company_id = str(company.id)
+
+    async def check_token_validity(self, request: Request, token: dict[str, Any]):
+        """
+        Check if the token is not expired or if it is not enabled, otherwise
+        it throws an AuthorizationError
+        """
+        now = utcnow()
+        if token.get("expiry") and token.get("expiry") < now:
+            raise AuthorizationError(403, gettext("Token expired."), title=gettext("403. Forbidden"))
+
+        if not token.get("enabled", False):
+            raise AuthorizationError(
+                403,
+                gettext("The requested resource is not available for your subscription."),
+                title=gettext("403. Forbidden"),
+            )
+
+    async def check_rate_limit(self, request: Request, token: dict[str, Any]):
+        updates = {}
+        now = utcnow()
+        new_period = False
+        rate_limit_requests = get_app_config("RATE_LIMIT_REQUESTS")
+        tokens_service = get_resource_service(API_TOKENS)
+
+        if rate_limit_requests:
+            new_period = not token.get("rate_limit_expiry") or token["rate_limit_expiry"] <= now
+            if new_period:
+                updates["rate_limit_requests"] = 1
+            elif token.get("rate_limit_expiry"):
+                if token.get("rate_limit_requests", 0) >= rate_limit_requests:
+                    await request.abort(429, gettext("Rate limit exceeded"))
+                else:
+                    updates["rate_limit_requests"] = token.get("rate_limit_requests", 0) + 1
+
+        rate_limit_period = get_app_config("RATE_LIMIT_PERIOD")
+        if rate_limit_period and new_period:
+            updates["rate_limit_expiry"] = now + timedelta(seconds=rate_limit_period)
+
+        if updates:
+            tokens_service.patch(token.get("token"), updates)
+
+            # TODO-ASYNC: check if we should use request.storage instead
```

Comment:
Yes, internally they point to the same thing. `request.storage.request` is a wrapper around `flask.g` and `request.storage.session` is a wrapper around `flask.session`

So you should be able to replace any code that uses the above with the new request storage equivalent.